scripts/extract-firrtl-code.lua

```diff
@@ -0,0 +1,76 @@
+index = 0
+
+-- Return the indices of all strings in `text` that are between `snippetbegin`
+-- and `snippetend` strings.
+function getSnippets (text)
+   local foundSnippet = false
+   local i = 0
+   local start
+   local snippets = {}
+   while true do
+      if start == nil then
+         _, i = string.find(text, "snippetbegin.-\n", i + 1)
+         if i == nil then
+            break
+         end
+         -- Increment by one to drop the newline from the snippet.
+         start = i + 1
+         -- Decrement 'i' so that there is _always_ a newline for the match in
+         -- the else block to use.  This makes it possible to match emtpy
+         -- snippets easily as they look exactly like non-empty snippets.
+         i = i - 1
+      else
+         e, i = string.find(text, "\n[^\n]-snippetend", i + 1)
+         if i == nil then
+            error("missing 'snippetend'")
+         end
+         table.insert(snippets, {i = start, e = e - 1})
+         start = nil
+      end
+   end
+   return snippets
+end
+
+function CodeBlock(elem)
+   local extMap = {
+      firrtl = "fir",
+      verilog = "v",
+      systemverilog = "sv"
+   }
+   local ext = extMap[elem.classes[1]]
+   local skip = extMap[elem.classes[2]] == notest
```

Comment:
This allows opting out from testing for a specific snippet using syntax like `{.firrtl .notest}` on the code block. This is very helpful for things that are really annoying to test like Verilog that includes input files or FIRRTL code blocks that are intentionally not legal through the use of pseudo-type parameters (see memories).